quickwit/quickwit-indexing/src/source/queue_sources/shared_state.rs

```diff
@@ -18,38 +18,104 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 use std::collections::BTreeMap;
+use std::sync::Arc;
 use std::time::Duration;
 
 use anyhow::{bail, Context};
 use quickwit_metastore::checkpoint::PartitionId;
 use quickwit_proto::metastore::{
     AcquireShardsRequest, MetastoreService, MetastoreServiceClient, OpenShardSubrequest,
-    OpenShardsRequest,
+    OpenShardsRequest, PruneShardsRequest,
 };
 use quickwit_proto::types::{DocMappingUid, IndexUid, Position, ShardId};
 use time::OffsetDateTime;
-use tracing::info;
+use tracing::{error, info};
 
 use super::message::PreProcessedMessage;
 
 #[derive(Clone)]
 pub struct QueueSharedState {
-    pub metastore: MetastoreServiceClient,
+    metastore: MetastoreServiceClient,
     pub index_uid: IndexUid,
     pub source_id: String,
     /// Duration after which the processing of a shard is considered stale and
     /// should be reacquired
-    pub reacquire_grace_period: Duration,
+    reacquire_grace_period: Duration,
+    _cleanup_handle: Arc<()>,
 }
 
 impl QueueSharedState {
+    /// Create a shared state service and runs a cleanup task that prunes shards
+    /// in the background
+    pub fn new(
+        metastore: MetastoreServiceClient,
+        index_uid: IndexUid,
+        source_id: String,
+        reacquire_grace_period: Duration,
+        max_age: Option<u32>,
+        max_count: Option<u32>,
+        pruning_interval: Duration,
+    ) -> Self {
+        let cleanup_handle = Arc::new(());
+        tokio::spawn(Self::run_cleanup_task(
```

Comment:
I don't understand. This runs on every node?

Reply:
Can you add some developer documentation in the source/queue_sources/design.md, or point me to some other place if that exists somewhere.